docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+.. under the License.
+
+.. _format_columnar:
+
+*********************
+Arrow Columnar Format
+*********************
+
+The "Arrow Columnar Format" includes a language-agnostic in-memory
+data structure specification, metadata serialization, and a protocol
+for serialization and generic data transport.
+
+This document is intended to provide adequate detail to create a new
+implementation of the columnar format without the aid of an existing
+implementation. We utilize Google's `Flatbuffers`_ project for
+metadata serialization, so it will be necessary to refer to the
+project's `Flatbuffers protocol definition files <FlatbuffersFiles>`_
+while reading this document.
+
+The columnar format has some key features:
+
+* Data adjacency for sequential access (scans)
+* O(1) (constant-time) random access
+* SIMD and vectorization-friendly
+* Relocatable without "pointer swizzling", allowing for true zero-copy
+  access in shared memory
+
+Some things to keep in mind:
+
+* Some data types can be mutated in-place in memory, but others
+  cannot. The community has made deliberate trade-offs regarding
+  in-place mutability to provide analytical performance and data
+  locality guarantees
+
+Terminology
+===========
+
+Since different projects have used different words to describe various
+concepts, here is a small glossary to help disambiguate.
+
+* **Array** or **Vector**: a sequence of values with known length all
+  having the same type. These terms are used interchangeably in
+  different Arrow implementations, but we use "array" in this
+  document.
+* **Slot**: a single logical value in an array of some particular data type
+* **Buffer** or **Contiguous memory region**: a sequential virtual
+  address space with a given length. Any byte can be reached via a
+  single pointer offset less than the region's length.
+* **Physical Layout type**: The underlying memory layout for an array
+  without taking into account any value semantics. For example, a
+  32-bit signed integer array and 32-bit floating point array have the
+  same layout type.
+* **Logical type**: An application-facing semantic value type that is
+  implemented using some physical layout type. For example, Decimal
+  values are stored as 16 bytes in a fixed-size binary
+  layout. Similarly, strings can be stored as ``List<1-byte>``. A
+  timestamp may be stored as 64-bit fixed-size layout.
+* **Nested** or **parametric type**: a data type whose full structure
+  depends on one or more other child types. Two fully-specified nested
+  types are equal if and only if their child types are equal. For
+  example, ``List<U>`` is distinct from ``List<V>`` iff U and V are
+  different types.
+* **Parent** and **child arrays**: names to express relationships
+  between physical value arrays in a nested type structure. For
+  example, a ``List<T>``-type parent array has a T-type array as its
+  child (see more on lists below).
+* **Leaf node** or **leaf**: A primitive value array that may or may
```

Comment:
"Primitive" (array or type) is not defined. Is it the contrary of "nested"?

Reply:
Additionally, this "leaf" term is not used anywhere in the document.

(to be honest, I also don't really understand the explanation here)